File: android/src/main/java/com/swmansion/rnscreens/ScreenWindowTraits.java
package com.swmansion.rnscreens;

import android.animation.ArgbEvaluator;
import android.animation.ValueAnimator;
import android.annotation.TargetApi;
import android.app.Activity;
import android.content.pm.ActivityInfo;
import android.os.Build;
import android.view.View;
import android.view.ViewParent;
import android.view.WindowInsets;
import android.view.WindowManager;

import androidx.annotation.Nullable;
import androidx.core.view.ViewCompat;

import com.facebook.react.bridge.GuardedRunnable;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.UiThreadUtil;

public class ScreenWindowTraits {
  private static boolean mDidSetOrientation = false;
  private static boolean mDidSetStatusBarAppearance = false;
  private static Integer mDefaultStatusBarColor = null;

  protected static void applyDidSetOrientation() {
    mDidSetOrientation = true;
  }

  public static boolean didSetOrientation() {
    return mDidSetOrientation;
  }

  protected static void applyDidSetStatusBarAppearance() {
    mDidSetStatusBarAppearance = true;
  }

  public static boolean didSetStatusBarAppearance() {
    return mDidSetStatusBarAppearance;
  }

  protected static void setOrientation(Screen screen, final Activity activity) {
    if (activity == null) {
      return;
    }

    final Integer orientation;

    if (screen != null && screen.getScreenOrientation() != null) {
      orientation = screen.getScreenOrientation();
    } else {
      orientation = ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED;
    }

    activity.setRequestedOrientation(orientation);
  }

  protected static void setColor(Screen screenForColor, Screen ScreenForAnimated, final Activity activity, ReactContext context) {
    if (activity == null || context == null || Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
      return;
    }

    if (mDefaultStatusBarColor == null) {
      mDefaultStatusBarColor = activity.getWindow().getStatusBarColor();
    }

    final Integer color;
    final boolean animated;

    if (screenForColor != null && screenForColor.getStatusBarColor() != null) {
      color = screenForColor.getStatusBarColor();
    } else {
      color = mDefaultStatusBarColor;
    }

    if (ScreenForAnimated != null && ScreenForAnimated.isStatusBarAnimated() != null) {
      animated = ScreenForAnimated.isStatusBarAnimated();
    } else {
      animated = false;
    }

    UiThreadUtil.runOnUiThread(
            new GuardedRunnable(context) {
              @TargetApi(Build.VERSION_CODES.LOLLIPOP)
              @Override
              public void runGuarded() {
                activity
                        .getWindow()
                        .addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
                int curColor = activity.getWindow().getStatusBarColor();
                ValueAnimator colorAnimation =
                        ValueAnimator.ofObject(new ArgbEvaluator(), curColor, color);
                colorAnimation.addUpdateListener(
                        new ValueAnimator.AnimatorUpdateListener() {
                          @Override
                          public void onAnimationUpdate(ValueAnimator animator) {
                            activity.getWindow().setStatusBarColor((Integer) animator.getAnimatedValue());
                          }
                        });

                if (animated) {
                  colorAnimation.setDuration(300).setStartDelay(0);
                } else {
                  colorAnimation.setDuration(0).setStartDelay(300);
                }
                colorAnimation.start();
              }
            });
  }

  protected static void setStyle(Screen screen, final Activity activity, ReactContext context) {
    if (activity == null || context == null) {
      return;
    }

    final String style;

    if (screen != null && screen.getStatusBarStyle() != null) {
      style = screen.getStatusBarStyle();
    } else {
      style = "light";
    }

    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
      UiThreadUtil.runOnUiThread(
              new Runnable() {
                @TargetApi(Build.VERSION_CODES.M)
                @Override
                public void run() {
                  View decorView = activity.getWindow().getDecorView();
                  int systemUiVisibilityFlags = decorView.getSystemUiVisibility();
                  if ("dark".equals(style)) {
                    systemUiVisibilityFlags |= View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR;
                  } else {
                    systemUiVisibilityFlags &= ~View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR;
                  }
                  decorView.setSystemUiVisibility(systemUiVisibilityFlags);
                }
              });
    }
  }

  protected static void setTranslucent(Screen screen, final Activity activity, ReactContext context) {
    if (activity == null || context == null) {
      return;
    }

    final boolean translucent;

    if (screen != null && screen.isStatusBarTranslucent() != null) {
      translucent = screen.isStatusBarTranslucent();
    } else {
      translucent = true;
    }

    UiThreadUtil.runOnUiThread(
            new GuardedRunnable(context) {
              @TargetApi(Build.VERSION_CODES.LOLLIPOP)
              @Override
              public void runGuarded() {
                // If the status bar is translucent hook into the window insets calculations
                // and consume all the top insets so no padding will be added under the status bar.
                View decorView = activity.getWindow().getDecorView();
                if (translucent) {
                  decorView.setOnApplyWindowInsetsListener(
                          new View.OnApplyWindowInsetsListener() {
                            @Override
                            public WindowInsets onApplyWindowInsets(View v, WindowInsets insets) {
                              WindowInsets defaultInsets = v.onApplyWindowInsets(insets);
                              return defaultInsets.replaceSystemWindowInsets(
                                      defaultInsets.getSystemWindowInsetLeft(),
                                      0,
                                      defaultInsets.getSystemWindowInsetRight(),
                                      defaultInsets.getSystemWindowInsetBottom());
                            }
                          });
                } else {
                  decorView.setOnApplyWindowInsetsListener(null);
                }

                ViewCompat.requestApplyInsets(decorView);
              }
            });
  }

  protected static void setHidden(Screen screen, final Activity activity) {
    if (activity == null) {
      return;
    }

    final boolean hidden;

    if (screen != null && screen.isStatusBarHidden() != null) {
      hidden = screen.isStatusBarHidden();
    } else {
      hidden = false;
    }

    UiThreadUtil.runOnUiThread(
            new Runnable() {
              @Override
              public void run() {
                if (hidden) {
                  activity.getWindow().addFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN);
                  activity.getWindow().clearFlags(WindowManager.LayoutParams.FLAG_FORCE_NOT_FULLSCREEN);
                } else {
                  activity.getWindow().addFlags(WindowManager.LayoutParams.FLAG_FORCE_NOT_FULLSCREEN);
                  activity.getWindow().clearFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN);
                }
              }
            });
  }

  protected static Screen findScreenForTrait(Screen screen, Screen.WindowTraits trait) {
    if (!hasChildScreenWithTraitSet(screen, trait)) {
      if (checkTraitForScreen(screen, trait)) {
        return screen;
      } else {
        // if this screen has no trait set and there is no child with trait set, we look for a parent that has the trait set
        return findParentWithTraitSet(screen, trait);
      }
    }
    return null;
  }

  private static @Nullable Screen findParentWithTraitSet(Screen screen, Screen.WindowTraits trait) {
    ViewParent parent = screen.getContainer();
    while (parent != null) {
      if (parent instanceof Screen) {
        if (checkTraitForScreen((Screen) parent, trait)) {
          return (Screen) parent;
        }
      }
      parent = parent.getParent();
    }
    return null;
  }

  protected static boolean hasChildScreenWithTraitSet(Screen screen, Screen.WindowTraits trait) {
    if (screen == null || screen.getFragment() == null) {
      return false;
    }
    for (ScreenContainer sc : screen.getFragment().getChildScreenContainers()) {
      // we check only the top screen for the trait
      Screen topScreen = sc.getTopScreen();
      if (topScreen != null && checkTraitForScreen(topScreen, trait)) {
        return true;
      }
      if (hasChildScreenWithTraitSet(topScreen, trait)) {
        return true;
      }
    }
    return false;
  }

  private static boolean checkTraitForScreen(Screen screen, Screen.WindowTraits trait) {
    switch (trait) {
      case ORIENTATION:
        return screen.getScreenOrientation() != null;
      case COLOR:
        return screen.getStatusBarColor() != null;
      case STYLE:
        return screen.getStatusBarStyle() != null;
      case TRANSLUCENT:
        return screen.isStatusBarTranslucent() != null;
      case HIDDEN:
        return screen.isStatusBarHidden() != null;
      case ANIMATED:
        return screen.isStatusBarAnimated() != null;
      default:
        throw new IllegalArgumentException("Wrong trait passed: " + trait);
    }
  }

}

File: android/src/main/java/com/swmansion/rnscreens/ScreenWindowTraits.kt
package com.swmansion.rnscreens

import android.app.Activity
import android.content.pm.ActivityInfo
import android.os.Build
import android.view.View
import android.view.ViewParent
import androidx.core.graphics.Insets
import androidx.core.view.OnApplyWindowInsetsListener
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat
import com.facebook.react.bridge.ReactContext
import com.facebook.react.bridge.UiThreadUtil
import com.swmansion.rnscreens.Screen.WindowTraits

object ScreenWindowTraits {
    // Methods concerning statusBar management were taken from `react-native`'s status bar module:
    // https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/modules/statusbar/StatusBarModule.java
    private var didSetOrientation = false
    private var didSetStatusBarAppearance = false
    private var didSetNavigationBarAppearance = false

    private var windowInsetsListener =
        object : OnApplyWindowInsetsListener {
            override fun onApplyWindowInsets(
                v: View,
                insets: WindowInsetsCompat,
            ): WindowInsetsCompat {
                val defaultInsets = ViewCompat.onApplyWindowInsets(v, insets)

                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                    val windowInsets =
                        defaultInsets.getInsets(WindowInsetsCompat.Type.statusBars())

                    return WindowInsetsCompat
                        .Builder()
                        .setInsets(
                            WindowInsetsCompat.Type.statusBars(),
                            Insets.of(
                                windowInsets.left,
                                0,
                                windowInsets.right,
                                windowInsets.bottom,
                            ),
                        ).build()
                } else {
                    return defaultInsets.replaceSystemWindowInsets(
                        defaultInsets.systemWindowInsetLeft,
                        0,
                        defaultInsets.systemWindowInsetRight,
                        defaultInsets.systemWindowInsetBottom,
                    )
                }
            }
        }

    internal fun applyDidSetOrientation() {
        didSetOrientation = true
    }

    internal fun applyDidSetStatusBarAppearance() {
        didSetStatusBarAppearance = true
    }

    internal fun applyDidSetNavigationBarAppearance() {
        didSetNavigationBarAppearance = true
    }

    internal fun setOrientation(
        screen: Screen,
        activity: Activity?,
    ) {
        if (activity == null) {
            return
        }
        val screenForOrientation = findScreenForTrait(screen, WindowTraits.ORIENTATION)
        val orientation = screenForOrientation?.screenOrientation ?: ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED
        activity.requestedOrientation = orientation
    }

    internal fun setStyle(
        screen: Screen,
        activity: Activity?,
        context: ReactContext?,
    ) {
        if (activity == null || context == null || Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            return
        }
        val screenForStyle = findScreenForTrait(screen, WindowTraits.STYLE)
        val style = screenForStyle?.statusBarStyle ?: "light"

        UiThreadUtil.runOnUiThread {
            val decorView = activity.window.decorView
            val window = activity.window
            val controller = WindowInsetsControllerCompat(window, decorView)

            controller.isAppearanceLightStatusBars = style == "dark"
        }
    }

    internal fun setHidden(
        screen: Screen,
        activity: Activity?,
    ) {
        if (activity == null) {
            return
        }
        val screenForHidden = findScreenForTrait(screen, WindowTraits.HIDDEN)
        val hidden = screenForHidden?.isStatusBarHidden ?: false
        val window = activity.window
        val controller = WindowInsetsControllerCompat(window, window.decorView)

        UiThreadUtil.runOnUiThread {
            if (hidden) {
                controller.hide(WindowInsetsCompat.Type.statusBars())
            } else {
                controller.show(WindowInsetsCompat.Type.statusBars())
            }
        }
    }

    internal fun setNavigationBarHidden(
        screen: Screen,
        activity: Activity?,
    ) {
        if (activity == null) {
            return
        }

        val window = activity.window

        val screenForNavBarHidden = findScreenForTrait(screen, WindowTraits.NAVIGATION_BAR_HIDDEN)
        val hidden = screenForNavBarHidden?.isNavigationBarHidden ?: false

        if (hidden) {
            WindowInsetsControllerCompat(window, window.decorView).let { controller ->
                controller.hide(WindowInsetsCompat.Type.navigationBars())
                controller.systemBarsBehavior =
                    WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            }
        } else {
            WindowInsetsControllerCompat(
                window,
                window.decorView,
            ).show(WindowInsetsCompat.Type.navigationBars())
        }
    }

    internal fun trySetWindowTraits(
        screen: Screen,
        activity: Activity?,
        context: ReactContext?,
    ) {
        if (didSetOrientation) {
            setOrientation(screen, activity)
        }
        if (didSetStatusBarAppearance) {
            setStyle(screen, activity, context)
            setHidden(screen, activity)
        }
        if (didSetNavigationBarAppearance) {
            setNavigationBarHidden(screen, activity)
        }
    }

    private fun findScreenForTrait(
        screen: Screen,
        trait: WindowTraits,
    ): Screen? {
        val childWithTrait = childScreenWithTraitSet(screen, trait)
        if (childWithTrait != null) {
            return childWithTrait
        }
        return if (checkTraitForScreen(screen, trait)) {
            screen
        } else {
            // if there is no child with trait set and this screen has no trait set, we look for a parent
            // that has the trait set
            findParentWithTraitSet(screen, trait)
        }
    }

    private fun findParentWithTraitSet(
        screen: Screen,
        trait: WindowTraits,
    ): Screen? {
        var parent: ViewParent? = screen.container
        while (parent != null) {
            if (parent is Screen) {
                if (checkTraitForScreen(parent, trait)) {
                    return parent
                }
            }
            parent = parent.parent
        }
        return null
    }

    private fun childScreenWithTraitSet(
        screen: Screen?,
        trait: WindowTraits,
    ): Screen? {
        screen?.fragmentWrapper?.let {
            for (sc in it.childScreenContainers) {
                // we check only the top screen for the trait
                val topScreen = sc.topScreen
                val child = childScreenWithTraitSet(topScreen, trait)
                if (child != null) {
                    return child
                }
                if (topScreen != null && checkTraitForScreen(topScreen, trait)) {
                    return topScreen
                }
            }
        }
        return null
    }

    private fun checkTraitForScreen(
        screen: Screen,
        trait: WindowTraits,
    ): Boolean =
        when (trait) {
            WindowTraits.ORIENTATION -> screen.screenOrientation != null
            WindowTraits.STYLE -> screen.statusBarStyle != null
            WindowTraits.HIDDEN -> screen.isStatusBarHidden != null
            WindowTraits.ANIMATED -> screen.isStatusBarAnimated != null
            WindowTraits.NAVIGATION_BAR_HIDDEN -> screen.isNavigationBarHidden != null
        }
}
